File: Mobile/lib/app/data/models/user_model/position_model/position_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:geolocator/geolocator.dart';
import 'package:json_annotation/json_annotation.dart';

import '../../../../common/utils/date_time.dart';

part 'position_model.g.dart';

@JsonSerializable()
class PositionModel {
  @JsonKey(name: 'latitude')
  final double latitude;
  @JsonKey(name: 'longitude')
  final double longitude;
  @JsonKey(name: 'address')
  final String? address;
  @JsonKey(name: "create_at", fromJson: DateTimeUtils.dateTimeFromJson)
  final DateTime? createAt;
  @JsonKey(
      name: "update_at",
      fromJson: DateTimeUtils.dateTimeFromJson,
      toJson: coordinatesToJson)
  final DateTime? updateAt;

  static String coordinatesToJson(DateTime? updatedAt) =>
      DateTime.now().toIso8601String();

  PositionModel({
    required this.longitude,
    required this.latitude,
    this.address,
    this.createAt,
    this.updateAt,
  }) : super();

  factory PositionModel.fromSnapShot(
      DocumentSnapshot<Map<String, dynamic>> snapshot) {
    return PositionModel.fromJson(snapshot.data() ?? {});
  }

  factory PositionModel.fromPosition(Position position) {
    return PositionModel.fromJson(position.toJson());
  }

  PositionModel copyWith(
          {String? address,
          double? longitude,
          double? latitude,
          DateTime? createAt}) =>
      PositionModel(
          longitude: longitude ?? this.longitude,
          latitude: latitude ?? this.latitude,
          address: address ?? this.address,
          createAt: createAt ?? this.createAt);

  factory PositionModel.fromJson(Map<String, dynamic> json) =>
      _$PositionModelFromJson(json);

  Map<String, dynamic> toJson() => _$PositionModelToJson(this);
}

File: BackendAPI/api/models/position_model.py
import json
from datetime import datetime

from utils import date_time_parser


class PositionModelEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.isoformat()
        elif isinstance(obj, PositionModel):
            return obj.__dict__
        else:
            return json.JSONEncoder.default(self, obj)


class PositionModel:
    def __init__(self, latitude: float, longitude: float, address: str, create_at: datetime, update_at: datetime):
        self.latitude = latitude
        self.longitude = longitude
        self.address = address
        self.create_at = create_at
        self.update_at = update_at

    @staticmethod
    def from_json(json_data):
        return PositionModel(
            latitude=json_data['latitude'],
            longitude=json_data['longitude'],
            address=json_data['address'],
            create_at=date_time_parser(json_data['create_at']) if 'create_at' in json_data else None,
            update_at=date_time_parser(json_data['update_at']) if 'update_at' in json_data else None,
        )

    def to_dict(self):
        return {
            "latitude": self.latitude,
            "longitude": self.longitude,
            "address": self.address,
            "create_at": self.create_at.isoformat() if self.create_at is not None else None,
            "update_at": self.update_at.isoformat() if self.update_at is not None else None,
        }
    def to_json(self):
        return json.dumps(self, cls=PositionModelEncoder)
    def pair(self):
        return self.latitude, self.longitude

    def __str__(self):
        return str(self.to_dict())
